File: 101-200/141-150/146-LRUCache/LRUCache.cpp
#include <unordered_map>

using namespace std;

class Node {
    Node *prev;
    Node *next;
    int key;
    int value;
public:
    Node(int _key, int _value): prev(nullptr), next(nullptr), key(_key), value(_value) {};
    int getKey();
    int getValue();
    friend class DoublyLinkedList;
};

int Node::getKey() {
    return key;
}

int Node::getValue() {
    return value;
}

class DoublyLinkedList {
    Node *head;
    Node *end;
    int length;
public:
    DoublyLinkedList() : head(nullptr), end(nullptr), length(0) {};
    void insertHead(Node *node);
    void remove(Node *node);
    void update(Node *node, int value);
    Node *getHead();
    Node *getEnd();
    int size();
};

void DoublyLinkedList::insertHead(Node *node) {
    if (head)
        head->prev = node;
    else
        end = node;
    
    node->prev = nullptr;
    node->next = head;
    head = node;
    ++length;
}

void DoublyLinkedList::remove(Node *node) {
    if (node->prev)
        node->prev->next = node->next;
    else {
        head = node->next;
        if (head)
            head->prev = nullptr;
    }

    if (node->next)
        node->next->prev = node->prev;
    else {
        end = node->prev;
        if (end)
            end->next = nullptr;
    }
    
    --length;
}

void DoublyLinkedList::update(Node *node, int value) {
    node->value = value;
}

Node *DoublyLinkedList::getHead() {
    return head;
}

Node *DoublyLinkedList::getEnd() {
    return end;
}

int DoublyLinkedList::size() {
    return length;
}

class LRUCache {
    int capacity;
    unordered_map<int, Node*> hashMap;
    DoublyLinkedList linkedList;
public:
    LRUCache(int _capacity): capacity(_capacity) {};
    
    int get(int key) {
        auto iter = hashMap.find(key);
        if (iter == hashMap.end())
            return -1;

        linkedList.remove(iter->second);
        linkedList.insertHead(iter->second);
        return iter->second->getValue();
    }
    
    void put(int key, int value) {
        auto iter = hashMap.find(key);
        if (iter != hashMap.end()) {
            linkedList.update(iter->second, value);
            linkedList.remove(iter->second);
            linkedList.insertHead(iter->second);
        }
        else {
            linkedList.insertHead(new Node(key, value));
            hashMap[key] = linkedList.getHead();
        }

        if (linkedList.size() == capacity + 1) {
            Node *node = linkedList.getEnd();
            linkedList.remove(node);
            hashMap.erase(node->getKey());
            delete node;
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 101-200/141-150/146-LRUCache/LRUCache.py
class DoublyLinkedListNode(object):
    def __init__(self, key, val, prev=None, next=None):
        self.key = key
        self.value = val
        self.prev = prev
        self.next = next

class DoublyLinkedList(object):
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def insert_head(self, node):
        if not self.head:
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head.prev = node
            self.head = node

        self.size += 1

    def remove(self, node):
        if self.size == 1:
            self.head = self.tail = None
        elif self.head == node:
            self.head = self.head.next
            self.head.prev = None
        elif self.tail == node:
            self.tail = self.tail.prev
            self.tail.next = None
        else:
            node.prev.next = node.next
            node.next.prev = node.prev

        self.size -= 1

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """

        self.doublyll = DoublyLinkedList()
        self.table = {}
        self.capacity = capacity

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.table:
            get_node = self.table[key]
            self.doublyll.remove(get_node)
            self.doublyll.insert_head(get_node)
            return get_node.value

        return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """

        if key not in self.table:
            if self.doublyll.size == self.capacity:
                delete_node = self.doublyll.tail
                self.doublyll.remove(delete_node)
                del self.table[delete_node.key]

            new_node = DoublyLinkedListNode(key, value)
            self.doublyll.insert_head(new_node)
            self.table[key] = new_node
        else:
            update_node = self.table[key]
            update_node.value = value
            self.doublyll.remove(update_node)
            self.doublyll.insert_head(update_node)



# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)